app/models/callback.rb

```diff
@@ -1,23 +1,28 @@
 class Callback < Base
   attribute :telephone, :string
-  # may need to DateTime.parse
   attribute :phone_call_scheduled_at, :string
 
   validates :telephone, length: { minimum: 5, too_short: "Telephone number is too short (minimum is 5 characters)" }, format: { with: /\A[0-9\s+]+\z/, message: "Enter a telephone number in the correct format" }
-  # possible caching issue
   validates :phone_call_scheduled_at, callbacks: { method: :get_callback_booking_quotas }
 
-  def self.options
-    quotas = ApiClient.get_callback_booking_quotas
-    grouped_quotas = quotas.group_by(&:day)
-    options_hash = Hash.new { |hash, key| hash[key] = [] }
-    grouped_quotas.each do |day, data|
-      data.each do |x|
-        gmt_start_slot = Time.zone.parse(x.start_at.to_s).strftime("%I:%M %P")
-        gmt_end_slot = Time.zone.parse(x.end_at.to_s).strftime("%I:%M %P")
-        options_hash[day] << [gmt_start_slot.to_s + " - " + gmt_end_slot.to_s, x.start_at]
+  class << self
+    def options
+      quotas = ApiClient.get_callback_booking_quotas
+      grouped_quotas = quotas.group_by(&:day)
+      options_hash = Hash.new { |hash, key| hash[key] = [] }
+      grouped_quotas.each do |day, data|
```

Comment:
This is quite busy/hard to read, would something like this work?

```
def options
  quotas_by_day = ApiClient.get_callback_booking_quotas.group_by(&:day)
  quotas_by_day.transform_values do |quotas|
    quotas.map { |quota| option(quota) }
  end

  today = Time.now.strftime("%A")
  quotas.delete(today)
end

def option(quota)
  start = Time.zone.parse(quota.start_at.to_s).strftime("%I:%M %P")
  end = Time.zone.parse(quota.end_at.to_s).strftime("%I:%M %P")

  options_hash[day] << ["#{start.to_s } - #{end.to_s}", quota.start_at]
end
```

Reply:
can't quite get that to work, quotas.delete(today) would return today, end is a reserved word, options_hash is not declared, have removed extra grouped quotas line, think it looks ok.

Reply:
oh I just have a typo, should be:

```
def options
  quotas_by_day = ApiClient.get_callback_booking_quotas.group_by(&:day)
  quotas_by_day.transform_values do |quotas|
    quotas.map { |quota| option(quota) }
  end

  today = Time.now.strftime("%A")
  quotas_by_day.delete(today)
  quotas_by_day
end
```

Reply:
Some of the complexity seems to be combining logic for what slots are available, with the presentation of times for those slots? Can the model have a more structured storage of the times (as DateTime objects)? And then use a helper method to map the values on the form in the view partial - eg

```ruby
f.select :my_attribute, convert_time_zones_of_quotas(f.object.options), :first, :last
```